src/hotspot/share/gc/epsilon/epsilonHeap.cpp

```diff
@@ -150,15 +152,15 @@ HeapWord* EpsilonHeap::allocate_work(size_t size) {
   size_t used = _space->used();
 
   // Allocation successful, update counters
-  {
+  if (Metaspace::initialized()) {
```

Comment:
Feels odd to protect this entire block. It handles generic allocations, not only the metaspace. (I suspect this is only relevant during the initial allocation at startup, but it still looks weird). Would moving a check to `EpsilonMonitoringSupport::update_counters()` and protecting `MetaspaceCounters::update_performance_counters();` there be enough?

Reply:
I can't move `if (Metaspace::initialized())` into `EpsilonMonitoringSupport::update_counters()` because the code in this function also updates the global variable `_last_counter_update`.

Reply:
Yes, I understand that sacrifices the logging accuracy a bit: the `_last_counter_update` would include the initial CDS allocation. At the same time, I would not like to put `Metaspace::initialized()` checks on this (hot-when-not-TLAB) path, only to catch that initial allocation. So I suggest we add a boolean parameter to `allocate_work(size_t size, bool verbose)`, and call it with `false` from `allocate_loaded_archive_space`.

Reply:
Fixed as you suggested in version  71c2aec.